### DISCUSSION TITLE
How to differ ELRS and CRSF in Lua

### DISCUSSION THREAD

**Discussion by alexey-gamov** [ExpressLRS/ExpressLRS#1302] [Q&A] [answered]
Hello! Some time ago started lua telemetry script for OpenTX monochrome radios.

Now i want to output power and frequency on LCD.

ELRS and CRSF have different RFMD values for freq if value is 1 or 2.

- ELRS: 1=25Hz, 2=50Hz
- CRSF: 1=50Hz, 2=150Hz

How i can detect what type is used in lua to show real freq?

**Comment by alexey-gamov** [comment]
Wrote test lua script from [this piece](https://github.com/ExpressLRS/ExpressLRS/blob/8cab8dbef316f8572d73a37518103c4b5502c4ef/src/lua/elrsV2.lua) - can't say if it's working on real hardware.
Is there more simple variant?

```lua
--[[
local function crossfireTelemetryPush()
	return true
end

local function crossfireTelemetryPop()
	-- dev: emulation crsf telemetry queue 
	return 0x29, {0x00, 0xEE, 0x00, 0x45, 0x4C, 0x52, 0x53}
end
]]

local function checkSerial(data, serial)
	local result = 0
	local offset = 3

	while data[offset] ~= 0 do
		offset = offset + 1
	end

	for i = 1, 4 do
		result = bit32.lshift(result, 8) + data[offset + i]
	end

	return (result == serial)
end

local function run(event)	
	if crsf then
		local command, data = crossfireTelemetryPop()

		if command == 0x29 and data[2] == 0xEE then
			elrs = checkSerial(data, 0x454C5253)
		end

		if elrs then
			popupWarning("You have ELRS!", event)
		else
			popupWarning("You have CRSF!", event)
		end
	else
		popupWarning("NO crossfire", event)
	end

	return 0
end

local function init()
	crsf = (crossfireTelemetryPush() ~= nil)
end

return { run = run, init = init }
```

**Reply by StonedDawg** [reply]
yes, check for serial number is how to differentiate them.

**Reply by alexey-gamov** [reply]
@StonedDawg Is there way to ask module for serial ones in init func?
So not to wait until telem pops..

**Reply by StonedDawg** [reply]
@alexey-gamov mmm... so you mean to catch the serial number on power up? no, we don't do them in init....but anyway by the time the you run the lua, the module has already enter the loop func?

**Comment by alexey-gamov** [comment]
Made it simpler, also added logs file. Now script should be placed in <kbd>TOOLS</kbd> folder. 
Already detect module name  `ES24TX Pro Ser`, but still can't get `ELRS` serial.

Help wanted.. Anyone?

```lua
local toolName = "TNS|ELRS Detector|TNE"

local function logger(result)
	local file = io.open("/SCRIPTS/TOOLS/detect.txt", "a")
	local timeNow = getDateTime()

	io.write(file, string.format("%02.0f:%02.0f:%02.0f - %s\n", timeNow.hour, timeNow.min, timeNow.sec, result))
	io.close(file)
end

local function run(event)
	if crsf and elrs == nil then
		local serial, command, packet = [[]], crossfireTelemetryPop()

		if command == 0x29 then
			logger(string.format("got command 0x%02X from 0x%02X", command, packet[2]))

			for i = 3, #packet - 1 do
				serial = serial .. string.char(packet[i])
			end

			logger(string.format("serial: %s", serial))

			if serial == "ELRS" then
				elrs = true
				logger(string.format("ELRS found! no more cycle", serial))
			end
		else
			crossfireTelemetryPush(0x28, {0x00, 0xEF})
		end
	end

	popupWarning(crsf and (elrs and "ELRS" or "CRSF") or "No crossfire", event)

	return 0
end

local function init()
	crsf = crossfireTelemetryPush() ~= nil
	logger(string.format("init() passed %s CRSF", crsf and "WITH" or "WITHOUT"))
end

return { run = run, init = init }
```

**Reply by alexey-gamov** [reply]
@StonedDawg, @CapnBry, can you assist a little bit?

**Reply by StonedDawg** [reply]
@alexey-gamov it's possible your logger might not print out the rest of the packet, because after "device name" there is a "null" before the Serial Number. processing your variable "serial" as a string would stop at the first null.

**Comment by CapnBry** [comment]
```
		local serial, command, packet = [[]], crossfireTelemetryPop() -- serial = [[]]??

		if command == 0x29 then
			logger(string.format("got command 0x%02X from 0x%02X", command, packet[2]))

                        -- the device name starts at 3, not the serial number
			for i = 3, #packet - 1 do
				serial = serial .. string.char(packet[i])
			end
...
                 crossfireTelemetryPush(0x28, {0x00, 0xEF})  -- should be 0xEA
```
I'm away, so short reply, but the serial number follows the device name, and you're init'ing a string to `[[]]` and spamming the wrong destination?
```
if command == 0x29 then
   local i = 3
  -- seek past the device name
  while packet[i] ~= 0 do
    i = i + 1
  end
  local isElrs = packet[i] == 0x45 and packet[i+1] == 0x4c and packet[i+2] == 0x52 and packet[i+3] == 0x53
```

Also, don't `crossfireTelemetryPush` every loop because it spams it continuously and every time you push this there's going to be the TX, the RX, the FC, and anything else responding, so only send it at most once every second. This is important because you're going to get 0x29 packets from EVERYTHING not just the TX module, so you should be prepared to have some things set isElrs == true and some will be isElrs == false (the flight controller) so you need to look at packet[2] (or whatever is the src address I can't remember) == 0xEE.

EDIT: Wait, I see your previous source was doing this all correctly? Why did you change it from what should have worked?

**Comment by alexey-gamov** [comment, marked as answer]
Now it works as it should. @StonedDawg & @CapnBry thank you for help!
```lua
local toolName = "TNS|ELRS Detector|TNE"

local function run(event)
	if crsf and elrs == nil then
		local shift, command, data = 3, crossfireTelemetryPop()

		if command == 0x29 and data[2] == 0xEE then
			while data[shift] ~= 0 do
				shift = shift + 1
			end

			elrs = (data[shift + 1] == 0x45 and data[shift + 2] == 0x4c and data[shift + 3] == 0x52 and data[shift + 4] == 0x53)
		else
			crossfireTelemetryPush(0x28, {0x00, 0xEA})
		end
	end

	popupWarning(crsf and (elrs and "ELRS" or "CRSF") or "No crossfire", event)

	return 0
end

local function init()
	crsf = crossfireTelemetryPush() ~= nil
end

return { run = run, init = init }
```

**Comment by CapnBry** [comment]
Note that if you use that for anything other than just a test, you:
* `local serial = [[]]` don't initialize serial to a pair of empty tables (should be a number, see below, or at worst a string `""`, not empty tables)
* `serial = serial .. string.cha(...)` don't convert serial to a string, just use it as a number so you don't create 4x strings for the garbage collector
* `crossfireTelemetryPush(0x28, {0x00, 0xEA})` this will spam like crazy since `run()` is called very frequently. The correct way is to send it once and wait for a timeout to send it again

All that is fine if you just want a tools script to tell you if the module is ELRS or CRSF, but if the ultimate goal is to use this in another script the above stuff should be fixed to prevent higher resource utilization.

**Reply by alexey-gamov** [reply]
Thx @CapnBry, made some changes in code above according to your advices.
Yes, there is some spam in `run()`, but it stops after detection and do not appear before next script start.
